src/transcript.rs

```diff
@@ -43,3 +44,85 @@ impl<F: Field> Hasher<F> for DummyHash<F> {
         tmp
     }
 }
+
+/// A transcript that can absorb points from both the base field and scalar
+/// field of a curve
+#[derive(Debug, Clone)]
+pub struct Transcript<C: CurveAffine, HBase, HScalar>
+where
+    HBase: Hasher<C::Base>,
+    HScalar: Hasher<C::Scalar>,
+{
+    // Hasher over the base field
+    base_hasher: HBase,
+    // Hasher over the scalar field
+    scalar_hasher: HScalar,
+    // Indicates if scalar(s) has been hashed but not squeezed
+    scalar_needs_squeezing: bool,
+    // PhantomData
+    _marker: PhantomData<C>,
+}
+
+impl<C: CurveAffine, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>
+    Transcript<C, HBase, HScalar>
+{
+    /// Initialise a new transcript with Field::one() as keys
+    /// in both the base_hasher and scalar_hasher
+    pub fn new() -> Self {
+        let base_hasher = HBase::init(C::Base::one());
+        let scalar_hasher = HScalar::init(C::Scalar::one());
+        Transcript {
+            base_hasher,
+            scalar_hasher,
+            scalar_needs_squeezing: false,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Initialise a new transcript with some given base_hasher and
+    /// scalar_hasher
+    pub fn init_with_hashers(base_hasher: &HBase, scalar_hasher: &HScalar) -> Self {
+        Transcript {
+            base_hasher: base_hasher.clone(),
+            scalar_hasher: scalar_hasher.clone(),
+            scalar_needs_squeezing: false,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Absorb a curve point into the transcript by absorbing
+    /// its x and y coordinates
+    pub fn absorb_point(&mut self, point: &C) -> Result<(), ()> {
+        let tmp = point.get_xy();
+        if bool::from(tmp.is_none()) {
+            return Err(());
+        };
+        let tmp = tmp.unwrap();
+        self.base_hasher.absorb(tmp.0);
+        self.base_hasher.absorb(tmp.1);
+        Ok(())
+    }
+
+    /// Absorb a base into the base_hasher
+    pub fn absorb_base(&mut self, base: C::Base) {
+        self.base_hasher.absorb(base);
+    }
+
+    /// Absorb a scalar into the scalar_hasher
+    pub fn absorb_scalar(&mut self, scalar: C::Scalar) {
+        self.scalar_hasher.absorb(scalar);
+        self.scalar_needs_squeezing = true;
+    }
+
+    /// Squeeze the transcript to obtain a C::Base value.
+    pub fn squeeze(&mut self) -> C::Base {
+        if self.scalar_needs_squeezing {
+            let transcript_scalar_point =
+                C::Base::from_bytes(&(self.scalar_hasher.squeeze()).to_bytes()).unwrap();
+            self.base_hasher.absorb(transcript_scalar_point);
```

Comment:
This correctly abstracts the existing mechanism for connecting the scalar and base field elements into a single transcript. Please open an issue for fixing this to have the properties we require.

Reply:
This could be related to https://github.com/zcash/halo2/issues/34.